### ISSUE TITLE
NER train on spacy v2

### ISSUE THREAD

**Issue by lioryaffe** [explosion/spaCy#1575]
Hi all,
so i have trained new entity using spacy 1.9 and everything was good.
I now want to upgrade to spacy v2, to see if my model is getting better results.
so I try to run the previous code but got errors, mainly with the
doc = nlp.make_doc(text)

so I try looking on examples, specially on: 
https://github.com/explosion/spaCy/blob/master/examples/training/train_new_entity_type.py

but my main issue is about the training data..
i have a folder with many files, each file represent a story. each file is tokenized so each word is in a separate line, so each line contains a word and a tag ('O' or 'MyNewEntity')

on the v1.9 i just used the:
doc = nlp.make_doc(text)
gold = GoldParse(doc, entities=tags) // where tags is a list of each word and the tag)
nlp.tagger(doc)
loss = nlp.entity.update(doc,gold)

I couldn't find the way to do the same thing in v2...
the example only shows how to pass the offsets inside a string, but my data is built differently

is there a way to do just minor changes in the oldest code for this to work?

## Your Environment
* Python Version Used: 3.6
* spaCy Version Used:  2.0

**Comment by ines** [comment]
> so I try to run the previous code but got errors

Which errors are you getting? 

You can definitely still pass in `GoldParse` objects the way you did before. The main changes to the training are around starting / ending the process, training the whole pipeline vs. individual components, and saving the models. So maybe you forgot to adjust something else in your code. You can find an overview of the most important changes to the training in v2 [in the migration guide](https://spacy.io/usage/v2#migrating-training).

The example scripts show the new and alternative "simple training style", which lets you pass in texts and a dictionaries of annotations (instead of `Doc` and `GoldParse` objects). It's slightly cleaner and less abstract, which is why we use it for the examples – but it's really just a question of personal preference. The [new training docs](https://spacy.io/usage/training#annotations) include more examples, as well as an explanation of the new "simple training style".

**Comment by lioryaffe** [comment]
My code (which is working with the 1.9 version):
```
nlp = spacy.load('en')
nlp.tokenizer = WhitespaceTokenizer

nlp.entity.add_label('ANIMAL')

for file in files:
    text = get_text(file)
    tags = get_tags(file)

    doc = nlp.make_doc(text)
    gold = GoldParse(doc, entities=tags)
    nlp.tagger(doc)
    loss = nlp.entity.update(doc, gold)
```

I'm getting an error on the make_doc function:
'str' object has no attribute 'vobac'

on the 1.9 version i loaded the model:
nlp = spacy.load('en', make_doc = WhitespaceTokenizer).
this got me error in v2, and after reading i made the adjustment with the nlp.tokenizer

**Comment by ines** [comment]
> `'str' object has no attribute 'vobac'`

How does your `WhitespaceTokenizer` look? Can you post the code? Since the error occurs in `make_doc`, which you've overwritten with a custom tokenizer, this is also where the error is. The most likely explanation is that somewhere in the `WhitespaceTokenizer`, you're trying to access the `.vocab` object of what's supposed to be a `Doc` – but it's actually the string of text.

**Comment by lioryaffe** [comment]
```
class WhitespaceTokenizer(object):
    def __init__(self, nlp):
        self.vocab = nlp.vocab

    def __call__(self, text):
        words = text.split(' ')
        spaces = [True] * len(words)
        return Doc(self.vocab, words=words, spaces=spaces)
```

so what is now instead of .vocab?

**Comment by ines** [comment]
Ahh – now it makes sense. The `WhitespaceTokenizer` is a class, which needs to be initialised with the `nlp` object, so it has access to the shared vocab. So you need to do the following:

```python
nlp.tokenizer = WhitespaceTokenizer(nlp)
```

This will initialise the tokenizer and set the tokenizer's vocab to the shared `nlp.vocab`. When you use the tokenizer on a string of text (in your case, via `nlp.make_doc(text)` – but `nlp.tokenizer(text)` works just as well), it will execute the `__call__` method of the custom tokenizer on that text.

**Comment by lioryaffe** [comment]
OK we are almost there :)

Now i'm getting 'spacy.gold.GoldParse' object is not iterable on the
loss = nlp.entity.update(doc, gold)

**Comment by ines** [comment]
The `update()` functions now take a *sequence* of `Doc` and `GoldParse` objects. This is useful, because in most cases on a larger scale, you would batch up your training data and train on batches of documents and annotations.

The easiest way in your code is to simply wrap them as a list. See the [training documentation](https://spacy.io/usage/training#annotations) for more details an examples – it's all explained there 😊

```python
nlp.entity.update([doc], [gold])
```

I'd also recommend checking out the [`train_new_entity_type.py`](https://github.com/explosion/spaCy/blob/master/examples/training/train_new_entity_type.py) example again and adjusting your code a little. Note how in spaCy v2.0, you now call `nlp.begin_training()`, which returns an optimizer to pass into your `update` function. You can also pass in an empty dictionary, which the loss(es) will be added to, and set a dropout rate to prevent the model from memorizing the examples.
